pkg/controller/csr/csr_controller_test.go

```diff
@@ -0,0 +1,330 @@
+		It("should reconcile the CSR controller", func() {
+			_, err = r.Reconcile(ctx, reconcile.Request{})
+			Expect(err).ShouldNot(HaveOccurred())
+		})
+
+		It("should reconcile a submitted CSR", func() {
+			Expect(cli.Create(ctx, validPod)).NotTo(HaveOccurred())
+			csr := newCSR(newX509CertificateRequest(), validPod)
+			Expect(cli.Create(ctx, csr)).NotTo(HaveOccurred())
+			_, err = r.Reconcile(ctx, reconcile.Request{})
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(r.client.Get(ctx, client.ObjectKey{Name: csr.Name}, csr)).NotTo(HaveOccurred())
+			Expect(csr.Status.Conditions).To(HaveLen(1))
+			Expect(csr.Status.Conditions[0].Type).To(Equal(certificatesv1.CertificateApproved))
+			Expect(csr.Status.Conditions[0].Status).To(Equal(corev1.ConditionTrue))
+			Expect(csr.Status.Certificate).ToNot(BeEmpty())
+		})
+
+		It("should reconcile 2 submitted CSRs", func() {
+			validPod2 := validPod.DeepCopy()
+			validPod2.Name = validPod2.Name + "2"
+			Expect(cli.Create(ctx, validPod2)).NotTo(HaveOccurred())
+			csr2 := newCSR(newX509CertificateRequest(), validPod2)
+			csr2.Name = csr2.Name + "2"
+			Expect(cli.Create(ctx, csr2)).NotTo(HaveOccurred())
+
+			Expect(cli.Create(ctx, validPod)).NotTo(HaveOccurred())
+			csr := newCSR(newX509CertificateRequest(), validPod)
+			Expect(cli.Create(ctx, csr)).NotTo(HaveOccurred())
+
+			_, err = r.Reconcile(ctx, reconcile.Request{})
+			Expect(err).ShouldNot(HaveOccurred())
+
+			Expect(r.client.Get(ctx, client.ObjectKey{Name: csr.Name}, csr)).NotTo(HaveOccurred())
+			Expect(csr.Status.Conditions).To(HaveLen(1))
+			Expect(csr.Status.Conditions[0].Type).To(Equal(certificatesv1.CertificateApproved))
+			Expect(csr.Status.Conditions[0].Status).To(Equal(corev1.ConditionTrue))
+			Expect(csr.Status.Certificate).ToNot(BeEmpty())
+
+			Expect(r.client.Get(ctx, client.ObjectKey{Name: csr2.Name}, csr2)).NotTo(HaveOccurred())
+			Expect(csr2.Status.Conditions).To(HaveLen(1))
+			Expect(csr2.Status.Conditions[0].Type).To(Equal(certificatesv1.CertificateApproved))
+			Expect(csr2.Status.Conditions[0].Status).To(Equal(corev1.ConditionTrue))
+			Expect(csr2.Status.Certificate).ToNot(BeEmpty())
+		})
+
+		It("should reject a submitted CSR that does not pass validation", func() {
+			csr := newCSR(newX509CertificateRequest(), validPod)
+			csr.Spec.Username = "attacker"
+			Expect(cli.Create(ctx, csr)).NotTo(HaveOccurred())
+			_, err = r.Reconcile(ctx, reconcile.Request{})
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(r.client.Get(ctx, client.ObjectKey{Name: csr.Name}, csr)).NotTo(HaveOccurred())
+			Expect(csr.Status.Conditions).To(HaveLen(1))
+			Expect(csr.Status.Conditions[0].Type).To(Equal(certificatesv1.CertificateDenied))
+			Expect(csr.Status.Conditions[0].Status).To(Equal(corev1.ConditionTrue))
+			Expect(csr.Status.Certificate).To(BeEmpty())
+		})
+	})
+
+	table.DescribeTable("csr validation", func(csrfunc func() *certificatesv1.CertificateSigningRequest, expectError bool) {
```

Comment:
I'm trying to follow this test and not sure why it takes a function that generates a CSR and not just a CSR object as input. Feels like an unnecessary layer here?